TideApp/TideApp/Managers/FavouritesManager.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  FavouritesManager.swift
+//  TideApp
+//
+//  Created by John Sanderson on 20/05/2021.
+//
+
+import Foundation
+import Combine
+import CoreLocation
+
+final class FavouritesManager {
+
+  // MARK: - Properties -
+  // MARK: Internal
+
+  @Published var favourites: [CLLocation] = []
+
+  // MARK: Private
+
+  private let userDefaults: UserDefaults
+  private var cancellable: AnyCancellable?
+
+  // MARK: - Initialiser -
+
+  init(userDefaults: UserDefaults = .standard) {
+    self.userDefaults = userDefaults
+  }
+
+  // MARK: - Functions -
+  // MARK: Internal
+
+  /// Tells the manager to start publishing the favourites
+  func start() {
+    cancellable = userDefaults
+      .publisher(for: \.favouriteLocations)
+      .decode(type: [Location].self, decoder: PropertyListDecoder())
+      .sink(receiveCompletion: { error in
+        self.favourites = []
+      }, receiveValue: { locations in
+        self.favourites = locations.map { CLLocation(latitude: $0.latitude, longitude: $0.longitude) }
+      })
+  }
+
+  /// Adds a new location to the favourites
+  /// - Parameter location: The new `CLLocation` to add
+  func add(_ location: CLLocation) {
+    encodeAndUpdate(with: location) { $0 + [$1] }
+  }
+
+  /// Removes the given location from the favourites
+  /// - Parameter location: The `CLLocation` to remove
+  func remove(_ location: CLLocation) {
+    encodeAndUpdate(with: location) { currentFavourites, proxy in
+      currentFavourites.filter { $0 != proxy }
+    }
+  }
+
+  // MARK: Private
+
+  private func encodeAndUpdate(with location: CLLocation, modifier: ([Location], Location) -> [Location]) {
```

Comment:
My 2 cents: 
this class seems to me over-complex, there is a .map called anytime you call encodeAndUpdate and a filter when you remove a location. I know we are not going to add a tons of locations but as an algorithm it takes really a lot of effort to just do something we can easily do using a dictionary.
So I would had a different approach were the manager process of adding and removing would be just set(value for key) and remove key. Basically it would be just an interface to access the UserDefault dictionary. This would also avoid to have same fav location because the name is right in the Location, and so avoid to use Set in the other parts (that is not necessary bad tbh)

eg: (I didnt put the combine stuff just for have a rapid example ⌛  )

```
struct Location: Codable {
  // This to be used all along our app, so mapping directly CLLocation or whatever in this
  let name: String
  let latitude: Double
  let longitude: Double
}

protocol FavouritesManagerType: AnyObject {
  func getFavouritesLocations() -> [Location] // or a var if preferred
  func saveToFavourite(_ location: Location)
  func removeFromFavourite(_ location: Location)
  func clearFavourites()
}

final class FavouritesManager: FavouritesManagerType {
  private static let favsLocationsKey: String = "FAVOURITES_LOCATIONS_STORAGE_KEY"
  private let storage: UserDefaults

  init(with storage: UserDefaults = .standard) { //tbh I don't like to provide a default for the dependency injection in general but its fine tho
    self.storage = storage
  }

  func getFavouritesLocations() -> [Location] {
    let locations = Array(retrieveLocationsDictionary().values)
    return locations
  }

  func saveToFavourite(_ location: Location) {
    var favsDictionary = retrieveLocationsDictionary()
    favsDictionary[location.name] = location
    updateLocationsDictionary(favsDictionary)
  }

  func removeFromFavourite(_ location: Location) {
    var favsDictionary = retrieveLocationsDictionary()
    favsDictionary[location.name] = nil
    updateLocationsDictionary(favsDictionary)
  }

  func clearFavourites() {
    storage.setValue(nil, forKey: FavouritesManager.favsLocationsKey)
  }

  private func updateLocationsDictionary(_ dictionary: [String: Location]) {
    let data = try? PropertyListEncoder().encode(dictionary)
    storage.setValue(data, forKey: FavouritesManager.favsLocationsKey)
  }

  private func retrieveLocationsDictionary() -> [String: Location] {
    guard let data = storage.data(forKey: FavouritesManager.favsLocationsKey) else { return [:] }
    return (try? PropertyListDecoder().decode([String: Location].self, from: data)) ?? [:]
  }
}
```

I repeat, my 2 cents about this, so happy to go with your solution 👍

Reply:
Sure, I see your point. To be honest, im not sure how this will play with combine... the nice thing about the current approach is that the `LocationsView` automatically responds to changes made to user defaults. 

I'll have a think

Reply:
Yeah, good point, well if it doesn't fit well, as said, happy with your approach!